File: longest-substring-without-repeating-characters.py
from collections import Counter

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        n = len(s)
        i = 0
        exists = Counter()
        mlen = 0
        for j in range(n):
            exists[s[j]] += 1
            while i < j and max(exists.values()) > 1:
                exists[s[i]] -= 1
                i += 1
            if max(exists.values()) <= 1:
                mlen = max(mlen, j - i + 1)
        return mlen

File: longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        int res = 0;
        for (int i = 0; i < n; i++) {
            int j = i;
            HashSet<Character> seen = new HashSet<>();
            while (j < n && !seen.contains(s.charAt(j))) {
                seen.add(s.charAt(j));
                j++;
            }
            res = Math.max(res, j - i);
        }
        return res;
    }
}